1.22 feature blog for minReadySeconds in StatefulSets (#28992)
kubernetes/website: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md

---
layout: blog

> title: 'minReadySeconds for StatefulSets'

Reviewer: Is there a typical format for the titles of post-release blog articles? If so, let's follow that.

Reviewer 2: I am not sure of that. Where do we have style guide?

Reviewer 2: @jlbutler ^?

Author: Apologies I thought I'd replied to this. As I mentioned last week, I also think this could use a more reader-friendly title. 

I don't know of any specific guideline but they do tend to be more descriptive of functionality, rather than calling out details.

> date: 2021-08-27

Reviewer: It's handy to update the path to the Markdown file to match. Hugo doesn't care but us humans can get confused.

slug: minreadyseconds-statefulsets
---

**Authors:** Ravi Gudimetla (Red Hat), Maciej Szulik (Red Hat)

This blog describes the notion of Availability for `StatefulSet` workloads, and a new alpha feature in Kubernetes 1.22 which adds `minReadySeconds` configuration for `StatefulSets`.

## What problems does this solve?

> Today, once a `StatefulSet` `Pod` is in the `Ready` state it is considered `Available` to receive traffic. For some of the `StatefulSet` workloads, it may not be the case. For example, a workload like Prometheus with multiple instances of Alertmanager, it should be considered `Available` only when Alertmanager's state transfer is complete, not when the `Pod` is in `Ready` state. Since `minReadySeconds` adds buffer, the state transfer may be complete before the `Pod` becomes `Available`. While this is not a fool proof way of identifying if the state transfer is complete or not, it gives a way to the end user to express their intention of waiting for sometime before the `Pod` is considered `Available` and it is ready to serve requests. 

Reviewer: Q. Does “today” describe the situation before the v1.22 release, or after it?

Reviewer 2: It is before v1.22

Reviewer: As this article goes out after the release, let's reword to make that clear.

> Prior to Kubernetes 1.22 release, once a `StatefulSet` `Pod` is in the `Ready` state it is considered `Available` to receive traffic. For some of the `StatefulSet` workloads, it may not be the case. For example, a workload like Prometheus with multiple instances of Alertmanager, it should be considered `Available` only when Alertmanager's state transfer is complete, not when the `Pod` is in `Ready` state. Since `minReadySeconds` adds buffer, the state transfer may be complete before the `Pod` becomes `Available`. While this is not a fool proof way of identifying if the state transfer is complete or not, it gives a way to the end user to express their intention of waiting for sometime before the `Pod` is considered `Available` and it is ready to serve requests. 

Reviewer 3: ```suggestion
Prior to Kubernetes 1.22 release, once a `StatefulSet` `Pod` is in the `Ready` state it is considered `Available` to receive traffic. For some of the `StatefulSet` workloads, that may not be the case. For example, a workload like Prometheus, with multiple instances of Alertmanager, should be considered `Available` only when Alertmanager's state transfer is complete, not when the `Pod` is in `Ready` state. Since `minReadySeconds` adds buffer, the state transfer may be complete before the `Pod` becomes `Available`. While this is not a foolproof way of identifying if the state transfer is complete or not, it gives a way to the end user to express their intention of waiting for some time before the `Pod` is considered `Available` and ready to serve requests. 
```

> Another case, where `minReadySeconds` helps is when using `LoadBalancer` `Services` with cloud providers. Since `minReadySeconds` adds latency after a `Pod` is `Ready`, it provides buffer time to prevent killing pods in rotation before new pods show up. Imagine a load balancer in unhappy path taking 10-15s to propagate. If you have 2 replicas then, you'd kill the second replica only after the first one is up but in reality, first replica cannot be seen because it is not yet ready to serve requests.

Reviewer 3: ```suggestion
Another case where `minReadySeconds` helps is when using `LoadBalancer` `Services` with cloud providers. Since `minReadySeconds` adds latency after a `Pod` is `Ready`, it provides buffer time to prevent killing pods in rotation before new pods show up. Imagine a load balancer in an unhappy path, taking 10-15s to propagate. If you have 2 replicas, then you'd kill the second replica only after the first one is up. But in reality, the first replica cannot be seen because it is not yet ready to serve requests.
```

> So, in general, the notion of `Availability` is `StatefulSets` is pretty useful and this feature helps in solving the above problems. This is a feature that already exists for `Deployments` and `DaemonSets` and we now have them for `StatefulSets` too to give users consistent workload experience.

Reviewer 4: ```suggestion
So, in general, the notion of `Availability` in `StatefulSets` is pretty useful and this feature helps in solving the above problems. This is a feature that already exists for `Deployments` and `DaemonSets` and we now have them for `StatefulSets` too to give users consistent workload experience.
```

Reviewer 2: Thank you @simonpasquier for the review

## How does it work?

> The `StatefulSet` controller watches for both `StatefulSets` and the `Pods` associated with them. When the feature gate associated with this feature is enabled, the `StatefulSet` controller identifies how long a particular `Pod` associated with a `StatefulSet` has been in the `Running` state.
> 
> If this value is greater than or equal to the time specified by the end user in `.spec.minReadySeconds` field, the `StatefulSet` controller updates a field called `AvailableReplicas` in the `StatefulSet`'s status subresource to include this `Pod`. The `AvailableReplicas` in `StatefulSet`'s status is a int32 field which tracks the number of pods that are `Available`.

Reviewer: tiny nit:
```suggestion
The statefulset controller watches for both `StatefulSets` and the `Pods` associated with them. When the feature gate associated with this feature is enabled, the statefulset controller identifies how long a particular `Pod` associated with a `StatefulSet` has been in the `Running` state.

If this value is greater than or equal to the time specified by the end user in `.spec.minReadySeconds` field, the statefulset controller updates a field called `availableReplicas` in the `StatefulSet`'s status subresource to include this `Pod`. The `.status.availableReplicas` field for a `StatefulSet` is an integer field that tracks the number of pods that are `Available`.
```

## How do I use it?

You are required to prepare the following things in order to try out the feature:

   - Download and install a kubectl greater than v1.22.0 version
   - Switch on the feature gate with the command line flag `--feature-gates=StatefulSetMinReadySeconds=true` on `kube-apiserver` and `kube-controller-manager`

After successfully starting `kube-apiserver` and `kube-controller-manager`, you will see `AvailableReplicas` in the status and `minReadySeconds` of spec (with a default value of 0). 

> Specify a value for `minReadySeconds` for any StatefulSet and you can check if `Pods` are available or not by checking `AvailableReplicas` field using:

Reviewer 3: ```suggestion
Specify a value for `minReadySeconds` for any StatefulSet and you can check if `Pods` are available or not by checking the `AvailableReplicas` field using:
```

`kubectl get statefulset/<name_of_the_statefulset> -o yaml`

## How can I learn more?

> KEP: https://github.com/kubernetes/enhancements/issues/2599
> API Changes: https://github.com/kubernetes/kubernetes/pull/100842
> Implementation PR: https://github.com/kubernetes/kubernetes/pull/101316

Reviewer: ```suggestion
- Read the KEP: [minReadySeconds for StatefulSets](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2599-minreadyseconds-for-statefulsets#readme)
- Read the documentation: [Minimum ready seconds](/docs/concepts/workloads/controllers/statefulset/#minimum-ready-seconds) for StatefulSet
- Review the [API definition](/docs/reference/kubernetes-api/workload-resources/stateful-set-v1/) for StatefulSet
```

## How do I get involved?

> Please reach out to us on [#sig-apps](https://kubernetes.slack.com/archives/C18NZM5K9) channel on slack, sig-apps mailing list-kubernetes-sig-apps@googlegroups.com

Reviewer: ```suggestion
Please reach out to us in the [#sig-apps](https://kubernetes.slack.com/archives/C18NZM5K9) channel on Slack (visit https://slack.k8s.io/ for an invitation if you need one), or on the SIG Apps mailing list: kubernetes-sig-apps@googlegroups.com
```
